### PR TITLE
upgrade pairing

### PR DESCRIPTION
**Pull request by kigawas**

### CODE REVIEW

**Comment by kigawas** [comment]
mock is not fixed yet, I'll do it tommorow

File: Cargo.toml
Comment (afck) [inline Cargo.toml]: Let's keep it in alphabetical order.
Also, whenever only a version is specified, let's use the simple form:
```toml
pairing = "0.15.0"
```

File: src/lib.rs
Comment (afck) [inline src/lib.rs]: Maybe `Vec::with_capacity(len)` to avoid reallocation. Or just one line:
```rust
iter::repeat_with(|| Fr::random(rng)).take(len).collect()
```

File: src/lib.rs
Comment (afck) [inline src/lib.rs]: Why not use `rng` itself?

File: src/lib.rs
Comment (afck) [inline src/lib.rs]: Why does
```rust
SecretKey(Box::new(Fr::random(rng)))
```
not work?
(I also wonder whether we can get rid of the `Box`, now that we're using `Zeroize`?)

**Review by afck — COMMENTED** [review]
Thanks! It's great to finally get rid of rand 0.4! :tada:

File: src/lib.rs
Comment (kigawas) [reply src/lib.rs]: Since `?Size` trait bound in `sample`'s signature, you can only use a `Sized` rng instead

File: src/lib.rs
Comment (kigawas) [reply src/lib.rs]: Not bad, but I hope rust can give me something like list comprehension :P

File: src/lib.rs
Comment (kigawas) [reply src/lib.rs]: Or we can add `sample` function ourselves, without implementing `Distribution`

Comment (kigawas) [reply src/lib.rs]:
```rust
    pub fn sample<R: RngCore>(&self, rng: &mut R) -> Self {
        Self(Fr::random(rng))
    }
```

File: src/lib.rs
Comment (kigawas) [inline src/lib.rs]: IMO, this looks neater than `Distribution`

File: src/lib.rs
Comment (kigawas) [inline src/lib.rs]: IMO, this looks neater than `Distribution`

File: src/lib.rs
Comment (kigawas) [inline src/lib.rs]: And we may remove this

**Comment by kigawas** [comment]
After https://github.com/zkcrypto/pairing/pull/108 and related prs merged we can solve the Sized rng problem

File: src/lib.rs
Comment (afck) [inline src/lib.rs]: Now I'm confused why we can't just copy `fr`. That should have the same effect, shouldn't it?
```rust
let sk = SecretKey(*fr);
clear_fr(fr);
sk
```
Do we still need this function at all?

File: src/lib.rs
Comment (afck) [inline src/lib.rs]: Good catch! :+1:

File: src/mock/mod.rs
Comment (afck) [inline src/mock/mod.rs]: Let's always put an empty line between methods.

**Review by afck — COMMENTED** [review]
Looks good, but I'm still confused about some of the `ChaChaRng`s.

**Comment by afck** [comment]
I noticed that our `README.md` has outdated information about `rand`. This part can probably be removed: ``with the exception of `rand`… ``

File: src/lib.rs
Comment (kigawas) [reply src/lib.rs]: Good catch

**Comment by kigawas** [comment]
> I noticed that our `README.md` has outdated information about `rand`. This part can probably be removed: `` with the exception of `rand`… ``

#94

File: src/lib.rs
Comment (afck) [reply src/lib.rs]: This is the only point that isn't clear to me yet: Is there a reason why we shouldn't just use `rng` itself here?

**Comment by kigawas** [comment]
After https://github.com/zkcrypto/group/pull/4

**Comment by metajack** [comment]
Any updates here?

**Comment by kigawas** [comment]
> Any updates here?

Maybe you can ask here https://github.com/zkcrypto/group/pull/4

**Comment by afck** [comment]
Feel free to update the Rust version in `.travis.yml` — maybe that's why parity-scale-codec doesn't compile?

**Comment by kigawas** [comment]
It seems group updated, though the pr not merged yet.

Other stuff LGTM, ready to merge

**Review by afck — APPROVED** [review]
Looks great, thank you! :+1:

**Comment by metajack** [comment]
Will a new version be published soon with this change?

**Comment by afck** [comment]
I think so, but let's wait until #101 is merged, too.

**Pull request merged**
